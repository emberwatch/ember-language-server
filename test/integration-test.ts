import * as cp from 'child_process';
import * as path from 'path';
import * as fs from 'fs';

import { createMessageConnection, MessageConnection, Logger, IPCMessageReader, IPCMessageWriter } from 'vscode-jsonrpc';
import {
  DidOpenTextDocumentNotification,
  InitializeRequest,
  CompletionRequest,
  DefinitionRequest,
  Definition,
  TextDocumentPositionParams
} from 'vscode-languageserver-protocol';

function startServer() {
  const serverPath = './lib/start-server.js';

  return cp.fork(serverPath, [], {
    cwd: path.join(__dirname, '..')
  });
}

function openFile(connection: MessageConnection, filePath: string) {
  connection.sendNotification(DidOpenTextDocumentNotification.type, {
    textDocument: {
      uri: `file://${filePath}`,
      text: fs.readFileSync(filePath, 'utf8')
    }
  });
}

function replaceDynamicUriPart(uri: string) {
  let dirname = __dirname;
  if (dirname.indexOf(':') === 1) {
    dirname = dirname.substr(2);
  }

  return uri
    .replace(dirname.replace(/\\/g, '/'), '/path-to-tests')
    .replace(dirname, '/path-to-tests')
    .replace(/\\/g, '/');
}

function normalizeUri(objects: Definition) {
  if (!Array.isArray(objects)) {
    objects.uri = replaceDynamicUriPart(objects.uri);
    return objects;
  }

  return objects.map(object => {
    if (object.uri) {
      const { uri } = object;
      object.uri = replaceDynamicUriPart(object.uri);
    }

    return object;
  });
}

function createParams(path: string, line: number, column: number): TextDocumentPositionParams {
  return {
    textDocument: {
      uri: `file://${path}`
    },
    position: {
      line,
      character: column
    }
  };
}

describe('integration', function() {
  let connection: MessageConnection;
  let serverProcess: cp.ChildProcess;

  async function getDefinition(path: string, line: number, column: number) {
    openFile(connection, path);

    let params = createParams(path, line, column);
    let definition = await connection.sendRequest(DefinitionRequest.type, params);

    return normalizeUri(definition);
  }

  async function getCompletion(path: string, line: number, column: number) {
    openFile(connection, path);

    let params = createParams(path, line, column);

    return await connection.sendRequest(CompletionRequest.type, params);
  }

  beforeAll(() => {
    serverProcess = startServer();
    connection = createMessageConnection(
      new IPCMessageReader(serverProcess),
      new IPCMessageWriter(serverProcess),
      <Logger>{
        error(msg) { console.log('error', msg); },
        log(msg) { console.log('log', msg); },
        info(msg) { console.log('info', msg); },
        warn(msg) { console.log('warn', msg); },
      }
    );

    connection.listen();
  });

  afterAll(() => {
    connection.dispose();
    serverProcess.kill();
  });

  describe('Classic App Structure', () => {
    // This test actually needs to be run first,
    // otherwise the project won't have been initialized and the following tests will fail.
    describe('Initialize request', () => {
      it('returns an initialize request', async () => {
        const params = {
          rootUri: `file://${path.join(__dirname, 'fixtures', 'full-project')}`,
          capabilities: {}
        };

        const response = await connection.sendRequest(InitializeRequest.type, params);

        expect(response).toMatchSnapshot();
      });
    });

    describe('Completion request', () => {
      it('returns all components and helpers when requesting completion items in a handlebars expression', async () => {
        const applicationTemplatePath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'templates', 'application.hbs');

        let response = await getCompletion(applicationTemplatePath, 1, 2);

        expect(response).toMatchSnapshot();
      });

      it('returns all routes when requesting completion items in an inline link-to', async () => {
        const templatePath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'templates', 'definition.hbs');

        let response = await getCompletion(templatePath, 2, 23);

        expect(response).toMatchSnapshot();
      });

      it('returns all routes when requesting completion items in a block link-to', async () => {
        const templatePath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'templates', 'definition.hbs');

        let response = await getCompletion(templatePath, 3, 12);

        expect(response).toMatchSnapshot();
      });
    });

    describe('Definition request', () => {
      it('returns the definition information for a component in a template', async () => {
        const definitionTemplatePath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'templates', 'definition.hbs');

        let response = await getDefinition(definitionTemplatePath, 0, 4);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a helper in a template', async () => {
        const definitionTemplatePath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'templates', 'definition.hbs');

        let response = await getDefinition(definitionTemplatePath, 1, 4);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a hasMany relationship', async () => {
        const modelPath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'models', 'model-a.js');

        let response = await getDefinition(modelPath, 4, 27);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a belongsTo relationship', async () => {
        const modelPath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'models', 'model-b.js');

        let response = await getDefinition(modelPath, 4, 27);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a transform', async () => {
        const modelPath = path.join(__dirname, 'fixtures', 'full-project', 'app', 'models', 'model-a.js');

        let response = await getDefinition(modelPath, 6, 27);

        expect(response).toMatchSnapshot();
      });
    });
  });

  describe('Pods App Structure', () => {
    let projectRoot = [__dirname, 'fixtures', 'pod-project'];
    let podRoot = [...projectRoot, 'app', 'pod-dir'];

    describe('Pods Initialize request', () => {
      it('returns an initialize request', async () => {
        const params = {
          rootUri: `file://${path.join(...projectRoot)}`,
          capabilities: {}
        };

        const response = await connection.sendRequest(InitializeRequest.type, params);

        expect(response).toMatchSnapshot();
      });
    });

    describe('Completion request', () => {
      it('returns all components and helpers when requesting completion items in a handlebars expression', async () => {
        const templatePath = path.join(...podRoot, 'foo', 'template.hbs');

        let response = await getCompletion(templatePath, 1, 2);

        expect(response).toMatchSnapshot();
      });

      it('returns all routes when requesting completion items in an inline link-to', async () => {
        const templatePath = path.join(...podRoot, 'components', 'test-component', 'template.hbs');

        let response = await getCompletion(templatePath, 2, 23);

        expect(response).toMatchSnapshot();
      });

      it('returns all routes when requesting completion items in a block link-to', async () => {
        const templatePath = path.join(...podRoot, 'components', 'test-component', 'template.hbs');

        let response = await getCompletion(templatePath, 3, 13);

        expect(response).toMatchSnapshot();
      });
    });

    describe('Definition request', () => {
      it('returns the definition information for a component in a template', async () => {
        const definitionTemplatePath = path.join(...podRoot, 'components', 'test-component', 'template.hbs');

        let response = await getDefinition(definitionTemplatePath, 0, 4);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a nested component in a template', async () => {
        const definitionTemplatePath = path.join(...podRoot, 'components', 'test-component', 'template.hbs');

        let response = await getDefinition(definitionTemplatePath, 1, 4);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a hasMany relationship', async () => {
        const modelPath = path.join(...podRoot, 'foo', 'model.js');

        let response = await getDefinition(modelPath, 4, 22);

        expect(response).toMatchSnapshot();
      });

      it('returns the definition information for a belongsTo relationship', async () => {
        const modelPath = path.join(...podRoot, 'bar', 'model.js');

        let response = await getDefinition(modelPath, 4, 24);

        expect(response).toMatchSnapshot();
      });
    });
  });
});
