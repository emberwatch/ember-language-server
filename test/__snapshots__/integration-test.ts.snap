// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration Classic App Structure Completion request returns all components and helpers when requesting completion items in a handlebars expression 1`] = `
Array [
  Object {
    "detail": "component",
    "kind": 7,
    "label": "another-awesome-component",
  },
  Object {
    "detail": "component",
    "kind": 7,
    "label": "my-awesome-component",
  },
  Object {
    "detail": "component",
    "kind": 7,
    "label": "nested/nested-component",
  },
  Object {
    "detail": "helper",
    "kind": 3,
    "label": "some-helper",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "action",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember 1.11.0",
    "kind": 3,
    "label": "component",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "1.11.0",
  },
  Object {
    "detail": "Ember 1.13.0",
    "kind": 3,
    "label": "concat",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "1.13.0",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "debugger",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember 2.1.0",
    "kind": 3,
    "label": "get",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "2.1.0",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "if",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "input",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "link-to",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "loc",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "log",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "mount",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "outlet",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "partial",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "render",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "textarea",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "unbound",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "unless",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
  },
]
`;

exports[`integration Classic App Structure Completion request returns all routes when requesting completion items in a block link-to 1`] = `
Array [
  Object {
    "detail": "route",
    "kind": 17,
    "label": "nested.nested-route",
  },
  Object {
    "detail": "route",
    "kind": 17,
    "label": "test-route",
  },
]
`;

exports[`integration Classic App Structure Completion request returns all routes when requesting completion items in an inline link-to 1`] = `
Array [
  Object {
    "detail": "route",
    "kind": 17,
    "label": "nested.nested-route",
  },
  Object {
    "detail": "route",
    "kind": 17,
    "label": "test-route",
  },
]
`;

exports[`integration Classic App Structure Definition request returns the definition information for a belongsTo relationship 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/full-project/app/models/model-a.js",
  },
]
`;

exports[`integration Classic App Structure Definition request returns the definition information for a component in a template 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/full-project/app/components/my-awesome-component.js",
  },
]
`;

exports[`integration Classic App Structure Definition request returns the definition information for a hasMany relationship 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/full-project/app/models/model-b.js",
  },
]
`;

exports[`integration Classic App Structure Definition request returns the definition information for a helper in a template 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/full-project/app/helpers/some-helper.js",
  },
]
`;

exports[`integration Classic App Structure Definition request returns the definition information for a transform 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/full-project/app/transforms/custom-transform.js",
  },
]
`;

exports[`integration Classic App Structure Initialize request returns an initialize request 1`] = `
Object {
  "capabilities": Object {
    "completionProvider": Object {
      "resolveProvider": true,
    },
    "definitionProvider": true,
    "documentSymbolProvider": true,
    "textDocumentSync": 1,
  },
}
`;

exports[`integration Pods App Structure Completion request returns all components and helpers when requesting completion items in a handlebars expression 1`] = `
Array [
  Object {
    "detail": "component",
    "kind": 7,
    "label": "test-component/nested-component",
  },
  Object {
    "detail": "component",
    "kind": 7,
    "label": "test-component",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "action",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember 1.11.0",
    "kind": 3,
    "label": "component",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "1.11.0",
  },
  Object {
    "detail": "Ember 1.13.0",
    "kind": 3,
    "label": "concat",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "1.13.0",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "debugger",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember 2.1.0",
    "kind": 3,
    "label": "get",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
    "version": "2.1.0",
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "if",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "input",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "link-to",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "loc",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "log",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "mount",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "outlet",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "partial",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "render",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 7,
    "label": "textarea",
    "usableIn": Array [
      "MustachePath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "unbound",
    "usableIn": Array [
      "MustachePath",
      "SubExpressionPath",
    ],
  },
  Object {
    "detail": "Ember",
    "kind": 3,
    "label": "unless",
    "usableIn": Array [
      "BlockPath",
      "MustachePath",
      "SubExpressionPath",
    ],
  },
]
`;

exports[`integration Pods App Structure Completion request returns all routes when requesting completion items in a block link-to 1`] = `
Array [
  Object {
    "detail": "route",
    "kind": 17,
    "label": "bar",
  },
  Object {
    "detail": "route",
    "kind": 17,
    "label": "foo.foo-nest",
  },
  Object {
    "detail": "route",
    "kind": 17,
    "label": "foo",
  },
]
`;

exports[`integration Pods App Structure Completion request returns all routes when requesting completion items in an inline link-to 1`] = `
Array [
  Object {
    "detail": "route",
    "kind": 17,
    "label": "bar",
  },
  Object {
    "detail": "route",
    "kind": 17,
    "label": "foo.foo-nest",
  },
  Object {
    "detail": "route",
    "kind": 17,
    "label": "foo",
  },
]
`;

exports[`integration Pods App Structure Definition request returns the definition information for a belongsTo relationship 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/pod-project/app/pod-dir/foo/model.js",
  },
]
`;

exports[`integration Pods App Structure Definition request returns the definition information for a component in a template 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/pod-project/app/pod-dir/components/test-component/component.js",
  },
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/pod-project/app/pod-dir/components/test-component/template.hbs",
  },
]
`;

exports[`integration Pods App Structure Definition request returns the definition information for a hasMany relationship 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/pod-project/app/pod-dir/bar/model.js",
  },
]
`;

exports[`integration Pods App Structure Definition request returns the definition information for a nested component in a template 1`] = `
Array [
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/pod-project/app/pod-dir/components/test-component/nested-component/component.js",
  },
  Object {
    "range": Object {
      "end": Object {
        "character": 0,
        "line": 0,
      },
      "start": Object {
        "character": 0,
        "line": 0,
      },
    },
    "uri": "file:///path-to-tests/fixtures/pod-project/app/pod-dir/components/test-component/nested-component/template.hbs",
  },
]
`;

exports[`integration Pods App Structure Pods Initialize request returns an initialize request 1`] = `
Object {
  "capabilities": Object {
    "completionProvider": Object {
      "resolveProvider": true,
    },
    "definitionProvider": true,
    "documentSymbolProvider": true,
    "textDocumentSync": 1,
  },
}
`;
